### ISSUE TITLE
Twisted Web's wsgi.input isn't sufficiently file-like in all cases

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4029]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4029|
|Type|defect|
|Created|2009-09-15 19:52:52Z|
|Branch|https://github.com/twisted/twisted/tree/wsgi-none-readline-4029|

The `read` method of the wsgi.input object sometimes raises a `TypeError`.  This seems to be because it can pass `None` to an underlying wrapped file-like object and some popular examples of such do not like to be passed `None`.  Specifically, `cStringIO.StringIO` (but not `StringIO.StringIO` and `file` raise `TypeError` when this is done.

<details><summary>Searchable metadata</summary>

```
trac-id__4029 4029
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_wsgi_none_readline_4029 branches/wsgi-none-readline-4029
branch_author__rikyu rikyu
status__closed closed
resolution__fixed fixed
component__web web
keywords__wsgi wsgi
time__1253044372000000 1253044372000000
changetime__1258408612000000 1258408612000000
version__None None
owner__ 
cc__rikyu
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27344]) Branching to 'wsgi.input-4029'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I've given this a small start, only on the testing side, in the branch above.  The idea here is to significantly improve the tests for wsgi.input behavior: currently, the test methods each cover too much, which is a stylistic problem I would like to remedy; additionally, they don't exercise the code with all type combinations (they only exercise the StringIO case - which works - not the file case - which does not), so I'm trying to make it easy to run all the tests against multiple file-like object types.

Unfortunately I don't think I'll finish this anytime soon.  If someone else wants to pick it up, please feel free.

**Comment by twisted-trac** [comment]
|<img alt="rikyu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|rikyu commented|
|-|-|

I missed this ticket before I created [#4113](https://github.com/twisted/twisted/issues/4113), which has an attached patch, although it's just a fix for this issue, not a fix to the aforementioned problems with the wsgi tests.

I'd really like my patch (or any patch to the readline issue) merged in, as this is a fairly significant issue for my project at this time. If there's anything I can do to expedite the process, please let me know.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I'll consider your patch on [#4113](https://github.com/twisted/twisted/issues/4113) as a resolution to this ticket.  If it looks okay, then we can resolve this and open another ticket for the improved test coverage which will tell us if any _other_ parts of `twisted.web.wsgi` has incorrect StringIO/file compatibility assumptions.

**Comment by twisted-trac** [comment]
|<img alt="rikyu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|rikyu commented|
|-|-|

Thanks very much. Let me know if there's anything else I can do.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

`/branches/wsgi.input-4029` became `/branches/wsgi.input-4114` to go with [#4114](https://github.com/twisted/twisted/issues/4114) where I'll finish what I started here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [27515]) Branching to 'wsgi-none-readline-4029'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [27516]) Apply wsgi_input_multipart.diff 

refs [#4029](https://github.com/twisted/twisted/issues/4029)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [27517]) Fix minor coding standard issues - remove superfluous parenthesis and use camelCase for the `useTempfile` parameter (and document it too)

refs [#4029](https://github.com/twisted/twisted/issues/4029)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

I fixed some trivial issues with the patch in the revision mentioned above.  Aside from those, this looks fine to me (there are still all the same issues with the test suite as I was originally trying to fix with this ticket, of course, but we'll deal with those later).   Assuming [buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/wsgi-none-readline-4029) agrees, I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [27518]) Merge wsgi-none-readline-4029

Author: rikyu
Reviewer: exarkun
Fixes: [#4029](https://github.com/twisted/twisted/issues/4029)
Refs: [#4114](https://github.com/twisted/twisted/issues/4114)

Add a test for WSGI requests which end up with a `tempfile.TemporaryFile`
for `Request.content` and fix the implementation to deal with this case
by not passing `None` for the length argument to `readline` (which temp
files do not accept, unlike StringIOs).  WSGI applications can now receive
requests with large bodies.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
